receiver/gitlabreceiver/traces_receiver.go

```diff
@@ -105,10 +133,146 @@ func (gtr *gitlabTracesReceiver) Shutdown(ctx context.Context) error {
+	if e.ObjectAttributes.FinishedAt == "" {
+		gtr.logger.Debug("pipeline missing finishedAt timestamp, skipping...",
+			zap.String("status", e.ObjectAttributes.Status))
+		w.WriteHeader(http.StatusNoContent)
+		return
+	}
+
+	// Process the pipeline based on its status
+	switch e.ObjectAttributes.Status {
+	case "running", "pending", "created", "waiting_for_resource", "preparing", "scheduled":
+		gtr.logger.Debug("pipeline not complete, skipping...",
+			zap.String("status", e.ObjectAttributes.Status))
+		w.WriteHeader(http.StatusNoContent)
+		return
+	case "success", "failed", "canceled", "skipped":
+		// above statuses are indicators of a completed pipeline, so we process them
+		break
+	default:
+		gtr.logger.Warn("unknown pipeline status, skipping...",
+			zap.String("status", e.ObjectAttributes.Status))
+		w.WriteHeader(http.StatusNoContent)
+		return
+	}
+
+	traces, err := gtr.handlePipeline(e)
+	if err != nil {
+		gtr.failBadReq(ctx, w, http.StatusInternalServerError, err, 0)
+		return
+	}
+
+	if traces.SpanCount() > 0 {
+		err = gtr.traceConsumer.ConsumeTraces(ctx, traces)
+		if err != nil {
+			gtr.failBadReq(ctx, w, http.StatusInternalServerError, err, 0)
+			return
+		}
+		w.WriteHeader(http.StatusOK)
+	} else {
+		w.WriteHeader(http.StatusNoContent)
+	}
+
+	gtr.obsrecv.EndTracesOp(ctx, metadata.Type.String(), traces.SpanCount(), nil)
+}
+
+func (gtr *gitlabTracesReceiver) validateReq(r *http.Request) (gitlab.EventType, error) {
+	if r.Method != http.MethodPost {
+		return "", errInvalidHTTPMethod
+	}
+
+	if gtr.cfg.WebHook.Secret != "" {
+		secret := r.Header.Get(defaultGitLabSecretTokenHeader)
+		if secret != gtr.cfg.WebHook.Secret {
+			return "", fmt.Errorf("%w: %s", errInvalidHeader, defaultGitLabSecretTokenHeader)
+		}
+	}
+
+	for key, value := range gtr.cfg.WebHook.RequiredHeaders {
+		if r.Header.Get(key) != string(value) {
+			return "", fmt.Errorf("%w: %s", errInvalidHeader, key)
+		}
+	}
+
+	eventType := gitlab.WebhookEventType(r)
+	if eventType == "" {
+		return "", fmt.Errorf("%w: %s", errMissingHeader, defaultGitLabEventHeader)
+	}
+	return eventType, nil
+}
+
+func (gtr *gitlabTracesReceiver) failBadReq(ctx context.Context,
+	w http.ResponseWriter,
+	httpStatusCode int,
+	err error,
+	spanCount int,
+) {
+	defer gtr.obsrecv.EndTracesOp(ctx, metadata.Type.String(), spanCount, err)
+
+	if err == nil {
+		w.WriteHeader(httpStatusCode)
+		return
+	}
+
+	jsonResp, marshalErr := json.Marshal(err.Error())
+	if marshalErr != nil {
+		gtr.logger.Warn("failed to marshall error to json", zap.Error(marshalErr))
+		w.WriteHeader(httpStatusCode)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(httpStatusCode)
+
+	if _, writeErr := w.Write(jsonResp); writeErr != nil {
+		gtr.logger.Warn("failed to write json response", zap.Error(writeErr))
+	}
+
+	gtr.logger.Debug(string(jsonResp), zap.Int("http_status_code", httpStatusCode), zap.Error(err))
```

Comment:
```suggestion
	gtr.logger.Debug(string(jsonResp), zap.Int(semconv.AttributeHttpStatusCode, httpStatusCode), zap.Error(err))
```

nit: I don't know if that's the exact syntax (so you'd have to check), but instead of typing out the string attributes, you can use semconv to provide the typed constants.